optuna_dashboard/ts/components/GraphParetoFront.tsx

```diff
@@ -165,6 +165,58 @@ const makeMarker = (
   }
 }
 
+const getIsDominatedTrialND = (normalizedValues: number[][]) => {
+  // Fallback for straight-forward pareto front algorithm (O(N^2) complexity).
+  const dominatedTrials: boolean[] = []
+  normalizedValues.forEach((values0: number[]) => {
+    const dominated = normalizedValues.some((values1: number[]) => {
+      if (values0.every((value0: number, k: number) => values1[k] === value0)) {
+        return false
+      }
+      return values0.every((value0: number, k: number) => values1[k] <= value0)
+    })
+    dominatedTrials.push(dominated)
+  })
+  return dominatedTrials
+}
+
+const getIsDominatedTrial2D = (normalizedValues: number[][]) => {
+  // Fast pareto front algorithm (O(N log N) complexity).
+  const sorted = normalizedValues
+    .map((values, i) => [values[0], values[1], i])
+    .sort((a, b) => a[0] - b[0])
+  let minValue1 = sorted[0][1]
+  const dominatedTrials: boolean[] = new Array(normalizedValues.length).fill(
+    true
+  )
+
+  sorted.forEach((values) => {
+    if (values[1] <= minValue1) {
+      dominatedTrials[values[2]] = false
+      minValue1 = values[1]
+    }
+  })
+  return dominatedTrials
+}
```

Comment:
Is it OK when the following `sorted`?
```
sorted = [
    [0, 1, 0],
    [0, 2, 1],
    [0, 0, 2],
]
```

I think then `dominatedTrials = [false, true, false]`, but all trials are not dominated.

Reply:
Thank you for pointing it out!
I'll change it as follows:

```suggestion
const getIsDominatedTrial2D = (normalizedValues: number[][]) => {
  // Fast pareto front algorithm (O(N log N) complexity).
  const sorted = normalizedValues
    .map((values, i) => [values[0], values[1], i])
    .sort()
  let minValue1 = sorted[0][1]
  const dominatedTrials: boolean[] = new Array(normalizedValues.length).fill(
    true
  )

  sorted.forEach((values) => {
    if (values[1] <= minValue1) {
      dominatedTrials[values[2]] = false
      minValue1 = values[1]
    }
  })
  return dominatedTrials
}
```
Then, `sorted` will be
```
sorted = [
    [0, 0, 2],
    [0, 1, 0],
    [0, 2, 1],
]
```
and [0, 0, 2] will be correctly flagged as undominated (other two are dominated).

Reply:
@HideakiImamura How about this?